### PR TITLE
Using native price estimator

### LINKED ISSUE #1615: Implement optimized native price estimation
#1582 collected some ideas and requirements how to optimize the native token price calculation. This issue reiterates once more the requirements and presents concrete implementation ideas to achieve them.

1. We need a component optimized to calculate a tokens' price in terms of the native token (ETH, xDai)
2. Make use of existing `PriceEstimators` by being compatible with the `PriceEstimating` trait
3. It should utilize a cache to keep latency low
4. This cache needs to stay up-to-date by doing maintenance periodically (similar to `solvable_orders`)
5. The code needs to be shareable with the planned `/auction` endpoint

### Idea
Use existing PriceEstimators to compose a second estimator hierarchy which is optimized for caching native token price queries. To be able to maintain the native price cache in the background, a new type will be implemented which holds a `Weak<CachingPriceEstimator>` that will refetch token prices periodically triggered by `ServiceMaintenance`.
This way the requirements can be met without introducing too much new code while still having the option to easily switch out the implementation later on.

### Roadmap
- [ ] Implement `NativePriceEstimating` trait for all types implementing `PriceEstimating`
- [ ] Refactor existing price estimator code a bit
  - [ ] Make `SantitizedPriceEstimator` store a `Arc<dyn PriceEstimating>` instead of `T: PriceEstimating` (this is needed to create a second price estimator hierarchy specialized for native token price estimation)
  - [ ] Add helper fn in `/orderbook/src/main.rs` which instantiates PriceEstimators based on `&[PriceEstimatorType]` (this makes it possible to create an entirely separate estimator hierarchy with minimal code duplication)
- [ ] Update `MinFeeCalculator::compute_unsubsidized_min_fee` to use optimized price estimator
  - [ ] Create new estimator hierarchy `SanitizedPE` -> `CachingPE` -> `CompetitionPE`
  - [ ] Store new estimator in `MinFeeCalculator`
  - [ ] Use new estimator in `compute_unsubsidized_min_fee`
- [ ] Maintain cache for native token prices
  - [ ] Implement `PriceCacheMaintainer` which implements `Maintaining` and manages a `CachingPE`
  - [ ] Instantiate `PriceCacheMaintainer` for the `CachingPE` inside the native price estimator hierarchy and pass that to `ServiceMaintenance`

### The new `NativePriceEstimating` trait
```rust
#[mockall::automock]
#[async_trait::async_trait]
pub trait NativePriceEstimating: Send + Sync {
    async fn native_price(
        &self,
        token: &H160,
        native_token: &H160,
        sell_amount: &U256,
    ) -> Result<f64, PriceEstimationError> {
        self.native_prices(std::slice::from_ref(token), native_token, sell_amount)
            .await
            .into_iter()
            .next()
            .unwrap()
    }

    /// Returns one result for each query.
    async fn native_prices(
        &self,
        tokens: &[H160],
        native_token: &H160,
        sell_amount: &U256,
    ) -> Vec<Result<f64, PriceEstimationError>>;
}

#[async_trait::async_trait]
impl<T> NativePriceEstimating for T
where
    T: PriceEstimating,
{
    async fn native_prices(
        &self,
        tokens: &[H160],
        native_token: &H160,
        sell_amount: &U256,
    ) -> Vec<Result<f64, PriceEstimationError>> {
        let native_token_queries: Vec<_> = tokens
            .iter()
            .map(|token| Query {
                sell_token: *token,
                buy_token: *native_token,
                in_amount: *sell_amount,
                kind: OrderKind::Buy,
            })
            .collect();

        let estimates = self.estimates(&native_token_queries).await;

        estimates
            .into_iter()
            .zip(native_token_queries.iter())
            .map(|(estimate, query)| {
                estimate.map(|estimate| estimate.price_in_sell_token_f64(query))
            })
            .collect()
    }
}
```

### Maintaining the native price cache
```rust
pub struct PriceCacheMaintainer {
    estimator: Weak<CachingPriceEstimator>,
}

impl PriceCacheMaintainer {
    pub fn new(estimator: &Arc<CachingPriceEstimator>) -> Self {
        Self {
            estimator: Arc::downgrade(estimator),
        }
    }
}

#[async_trait::async_trait]
impl Maintaining for PriceCacheMaintainer {
    async fn run_maintenance(&self) -> anyhow::Result<()> {
        let estimator = match self.estimator.upgrade() {
            Some(estimator) => estimator,
            None => return Err(anyhow::anyhow!("estimator already freed")),
        };

        let queries: Vec<_> = estimator
            .cache
            .lock()
            .unwrap()
            .key_order()
            .cloned()
            .collect();
        let estimates = estimator.estimates(&queries).await;
        let queries_and_estimates = queries.iter().zip(estimates.into_iter());
        {
            let mut cache = estimator.cache.lock().unwrap();
            for (query, estimate) in queries_and_estimates {
                if let Ok(estimate) = estimate {
                    cache.cache_set(*query, estimate);
                }
            }
        }
        Ok(())
    }
}
```

### PR DESCRIPTION
**Pull request by MartinquaXD**
Fixes #1615 

It's time to actually create a new estimator hierarchy for native price estimation.
Because we now have a component specialized to cache native prices, it no longer makes that much sense to cache all the base estimators so `CachingPriceEstimator` and its metrics code can be dropped.
Dropping the `CachingPriceEstimator` means that it's now very straight forward to instantiate a price estimator for each passed price estimator type.

Except for dropping the cache the original price estimator hierarchy didn't change:
`Sanitized(Competition(Vec<Instrumented(BaseEstimator)>))`

The new native price estimation hierarchy is slightly different:
`Cache(NativePrice(Sanitized(Competition(Vec<Instrumented(BaseEstimator)>))))`

Which base estimators get instantiated for each estimator hierarchy can be configured with CLI arguments.
`--price-estimators` and `--native-price-estimators`.

We now also spawn a background task which maintains the native price cache. Right now only 3 parameters make sense for the particular maintenance strategy from which 2 can be configured by CLI arguments:
1. how long prices get cache (`--native-price-cache-max-age-secs`)
2. how many prices to update per maintenance cycle (`--native-price-cache-max-update-size`)
3. how often to trigger a maintenance cycle (currently hard coded to 1/second)

### Test Plan
Manual test

### CODE REVIEW

**Review by nlordell — APPROVED** [review]
Woohoo! Looks good to me!

File: crates/orderbook/src/main.rs [hunk 1]
```diff
@@ -187,17 +187,21 @@ struct Arguments {
...
 
-    /// The maximum number of price estimates that will be cached.
-    #[clap(long, default_value = "1000")]
-    price_estimator_cache_size: usize,
+    /// How many cached native token prices can be updated at most in one maintenance cycle.
+    #[clap(long, default_value = "3")]
+    native_price_cache_max_update_size: usize,
+
+    /// Which estimators to use to estimate token prices in terms of the chain's native token.
+    #[clap(long, env, default_value = "Baseline", arg_enum, use_delimiter = true)]
```
[the remark this answers is not in the document]
Comment (MartinquaXD) [reply crates/orderbook/src/main.rs:203]: Yes, absolutely! I will create the corresponding merge request for the pods and will merge them right after each other.

File: crates/orderbook/src/metrics.rs [hunk 2]
```diff
@@ -23,7 +23,6 @@ pub struct Metrics {
     /// Gas estimate metrics
     gas_price: Gauge,
     price_estimates: IntCounterVec,
-    price_estimator_cache: IntCounterVec,
```
Comment (vkgnosis) [inline crates/orderbook/src/metrics.rs:26]: Would be nice to keep these metrics for the native cache which iirc doesn't have any. Can be in separate PR.

**Review by vkgnosis — APPROVED** [review]
